src/mds/MDSRankQuiesceInterface.cc

```diff
@@ -275,24 +299,29 @@ void MDSRank::quiesce_agent_setup() {
             }
           }
         });
-        timer.add_event_after(quiesce_after, quiesce_task);
+        timer.add_event_after(quiesce_dummy, quiesce_task);
       }
+      return it->second.first;
     }
-
-    return it->second.first;
   };
 
   ci.cancel_request = [this, quiesce_requests](RequestHandle h) {
     std::lock_guard l(mds_lock);
     
     for (auto it = quiesce_requests->cbegin(); it != quiesce_requests->cend(); it++) {
-      if (it->second.first == h) {
-        if (it->second.second) {
-          it->second.second->complete(-ECANCELED);
+      if (auto reqid = it->second.first; reqid == h) {
+        if (mdcache->have_request(reqid)) {
```

Comment:
I'd expect this to be outside of the loop. The `h` is your `reqid`. Also, this will probably fail if no quiesce requests are added to the map - should we decide to not do this for the production path (i mean line 270 above)

Reply:
to clarify: I'm not concerned with the way it's written semantically, I appreciate that the code now combines the lookup for the debug quiesce case and the real deal and it does so effectively.

My concern is that it's a potential false positive when we cancel a request successfully that was supposed to be in the mdcache, but it wasn't, because we found it in the local map.

Maybe we should mark the request in the local map as such that must have been in the cache. Then we begin by checking with the cache, and if it says there's no such request then we proceed to searching for it in the local map. If we find it, we must validate that it was indeed a debug-only route.

Please note that another option is not to include a real request in the local debug request map on line 270. In that case we still need to first check the cache and then check locally and the we'll naturally avoid false positives

Reply:
> to clarify: I'm not simply concerned with the way it's written, I appreciate that the code now combines the lookup for the debug quiesce case and the real deal and it does so effectively.
> 
> My concern is that it's a potential false positive when we cancel a request that was supposed to be in the mdcache, but it wasn't, however, we still cancel successfully because we found it in the local map.

Consider: the quiesce db is "broken" in some way and a quiesce is left outstanding for a long period of time. The operator comes in and manually kills the quiesce op. Then at some point the db recovers and tries to cancel the request but it's already gone.

> Maybe we should mark the request in the local map as such that must have been in the cache. Then we begin by checking with the cache, and if it says there's no such request then we proceed to searching for it in the local map. If we find it, we must validate that it was indeed a debug-only route.
> 
> Please note that another option is not to include a real request in the local debug request map on line 270. In that case we still need to first check the cache and then check locally and the we'll naturally avoid false positives

I've moved the request_kill bit to the beginning so that the request is killed (if it is active) no matter if it's in the `quiesce_requests` map or not. However this is an odd aspect that the request may be killed (because it is indeed active) but it does not exist in the quiesce_requests map so the `cancel_request` method returns `ENOENT`.

Reply:
On mobile, haven't yet looked at the latest code. I'll try to address your comment regarding the ENOENT 

From the database perspective, the no entity response is one of the important ways to detect the expired condition when the set state is releasing. The call to cancel request must only return zero, i.e. success, when it correctly found the request by the handle and was able to cancel it. By receiving 0 as a response to cancel the agent assumes that the pause has been active until now, which is what allows for the guarantees behind the releasing-released transition. If there was no active request found at the time of the call, it should be reported as an error, and today any error from the cancel callback will result in the expired status on the root propagated to the expired status on the set.

Reply:
> Then at some point the db recovers and tries to cancel the request but it's already gone

If the cancelation is due to the db set state CANCELED or EXPIRED then there's no difference what the cancel method returns. The main consumer of the return code is the RELEASING state where it should only report RELEASED if the cancel aborted an active pause. In all other cases, the cancel should return a nonzero result and the agent will report EXPIRED to the db

Reply:
Please see: https://github.com/ceph/ceph/pull/54581/commits/74e6291e91c81eb2db8289dfebabad7fb84971e7

Let me know if I should squash that.